package/setup.py

```diff
@@ -130,33 +129,93 @@ def get(self, option_name, default=None):
         except configparser.NoOptionError:
             return default
 
-class MDAExtension(Extension, object):
-    """Derived class to cleanly handle setup-time (numpy) dependencies.
+class MDADistribution(Distribution, object):
+    """Derived class to cleanly handle setup-time (cython/numpy) dependencies.
     """
-    # The only setup-time numpy dependency comes when setting up its
-    #  include dir.
-    # The actual numpy import and call can be delayed until after pip
-    #  has figured it must install numpy.
-    # This is accomplished by passing the get_numpy_include function
-    #  as one of the include_dirs. This derived Extension class takes
-    #  care of calling it when needed.
-    def __init__(self, *args, **kwargs):
-        self._mda_include_dirs = []
-        super(MDAExtension, self).__init__(*args, **kwargs)
-
-    @property
-    def include_dirs(self):
-        if not self._mda_include_dirs:
-            for item in self._mda_include_dir_args:
-                try:
-                    self._mda_include_dirs.append(item()) #The numpy callable
-                except TypeError:
-                    self._mda_include_dirs.append(item)
-        return self._mda_include_dirs
-
-    @include_dirs.setter
-    def include_dirs(self, val):
-        self._mda_include_dir_args = val
+    # Setup-time cython dependency comes when cythonizing sources (cython might
+    #  not be available) or configuring numpy's include dir (numpy might
+    #  not be available either).
+    # The actual cython/numpy imports and calls can be delayed until after
+    #  pip/setuptools have figured they must install them.
+    # This is accomplished by using a distutils.core.Distribution subclass
+    #  that lazily fills-in cython and numpy information by hooking into
+    #  the run_commands method.
+    def __init__(self, attrs=None):
+        # Attributes must exist if they're to be imported from attrs.
+        self.mda_use_cython = is_dev
+        self.mda_build_requires = None
+        super(MDADistribution, self).__init__(attrs)
+
+    def run_commands(self):
+        # The following seems fragile...
```

Comment:
The rationale here is to commit to installing dependencies according to which class of setup command is invoked.
- `./setup.py install` is treated as an `install` and dependencies are installed;
- `./setup.py build` is treated as a `build` and an error is thrown if dependencies are needed (not installed automatically because there's no safe way to install them system-wide);
- `pip` first internally runs  `./setup.py egg_info` just to get dependency information. Normally we wouldn't want this to trigger dependency installs because `pip` would be on top of it, but `biopython` needs it to be done. Actual cythonization is done later, when `pip` internally invokes `./setup.py install`.

As you can see, the decisions are based on whether specific strings are present in the command (check `./setup.py --help-commands`). It'll fail in the future when someone comes up with a `whatevs_instl` that should behave like an `install`, or if `pip` starts using a different command than `egg_info` to list dependencies.